File: tools/binding_compliance/conformance/families/yaml_update_operations.py
"""YAML update consent guards and byte-authenticated owned rollback effects."""

import re
from collections.abc import Mapping
from functools import partial

from ..coverage import CoveragePredicate, FamilyCoveragePolicy

_CACHE = "cache/CLASSIC/yaml-cache/"
_MANIFEST_FILES = {_CACHE + "manifest-latest.json", _CACHE + "manifest.etag"}


def _files(value):
    """Accept a complete sorted file inventory with exact bytes or parsed manifest JSON."""
    if not isinstance(value, list) or any(
        not isinstance(item, Mapping)
        or set(item) not in ({"path", "hex"}, {"path", "json"})
        for item in value
    ):
        return None
    paths = [item["path"] for item in value]
    if any(
        not isinstance(path, str)
        or (path != ".env" and not path.startswith("cache/"))
        or ".." in path.split("/")
        for path in paths
    ) or paths != sorted(set(paths)):
        return None
    for item in value:
        if "hex" in item:
            if (
                not isinstance(item["hex"], str)
                or re.fullmatch(r"(?:[0-9a-f]{2})*", item["hex"]) is None
            ):
                return None
        elif item["path"] != _CACHE + "manifest-latest.json" or not isinstance(
            item["json"], Mapping
        ):
            return None
    return {item["path"]: item for item in value}


def _base(observation):
    """Require both independently read file checkpoints and actual native result records."""
    return (
        set(observation) == {"results", "beforeFiles", "files"}
        and isinstance(observation["results"], list)
        and all(isinstance(result, Mapping) for result in observation["results"])
        and _files(observation["beforeFiles"]) is not None
        and _files(observation["files"]) is not None
    )


def _payloads_unchanged(observation):
    """Manifest caching may change metadata; consent guards may not install YAML payloads."""
    before = _files(observation["beforeFiles"])
    after = _files(observation["files"])
    return {
        path: value for path, value in before.items() if path not in _MANIFEST_FILES
    } == {path: value for path, value in after.items() if path not in _MANIFEST_FILES}


def matches_check(api, observation):
    """Require an explicit disabled or available status without payload mutation."""
    if not _base(observation) or not _payloads_unchanged(observation):
        return False
    return any(
        result.get("api") == api
        and set(result)
        == {
            "api",
            "tag",
            "releaseTag",
            "publishedAt",
            "compatible",
            "incompatible",
            "incompatibleReasons",
            "unknownReason",
            "error",
        }
        and result["error"] is None
        and result["tag"] in (0, 1)
        and result["incompatible"] == []
        and result["incompatibleReasons"] == []
        and result["unknownReason"] == ""
        and (
            (
                result["tag"] == 0
                and result["compatible"] == []
                and result["releaseTag"] == ""
                and observation["beforeFiles"] == observation["files"]
            )
            or (
                result["tag"] == 1
                and bool(result["compatible"])
                and bool(result["releaseTag"])
            )
        )
        for result in observation["results"]
    )


def matches_apply(api, observation):
    """Require empty-consent success or a native refusal, with every payload unchanged."""
    if not _base(observation) or not _payloads_unchanged(observation):
        return False
    return any(
        result.get("api") == api
        and set(result) == {"api", "installed", "failed", "error"}
        and result["installed"] == []
        and result["failed"] == []
        and result["error"]
        in (None, "disabled", "malformed_approval", "stale_decision")
        for result in observation["results"]
    )


def matches_rollback(api, observation):
    """Authenticate restored generations rather than trusting a rolled-back flag."""
    if not _base(observation):
        return False
    before, after = _files(observation["beforeFiles"]), _files(observation["files"])
    for result in observation["results"]:
        if result.get("api") != api:
            continue
        if api == "yaml_rollback_update":
            if set(result) != {"api", "fileName", "rolledBack", "error"}:
                continue
            if result["error"] in ("invalid_name", "local_ignore_refused"):
                return before == after and result["rolledBack"] is False
            names, absent = (
                ([result["fileName"]], [])
                if result["rolledBack"] is True
                else ([], [result["fileName"]])
            )
            if result["error"] is not None:
                continue
        else:
            if (
                set(result)
                != {"api", "rolledBack", "noPrevious", "failedFiles", "failureReasons"}
                or result["failedFiles"] != []
                or result["failureReasons"] != []
            ):
                continue
            names, absent = result["rolledBack"], result["noPrevious"]
        valid = True
        for name in names:
            target = _CACHE + name
            previous = before.get(target + ".prev")
            current = before.get(target)
            restored = after.get(target)
            valid = (
                valid
                and previous is not None
                and restored is not None
                and "hex" in previous
                and "hex" in restored
                and restored["hex"] == previous["hex"]
            )
            if current is not None:
                valid = valid and after.get(target + ".prev", {}).get(
                    "hex"
                ) == current.get("hex")
        for name in absent:
            target = _CACHE + name
            valid = (
                valid
                and target + ".prev" not in before
                and before.get(target) == after.get(target)
            )
        if valid:
            return True
    return False


YAML_UPDATE_OPERATIONS_COVERAGE_POLICY = FamilyCoveragePolicy(
    "yaml-update-operations",
    tuple(
        CoveragePredicate(
            "yaml-update-operations." + suffix,
            "yaml-update-operations.execute",
            "yaml-update-operations.execute",
            family,
            (symbol,),
            partial(matcher, api),
            runtime_operations=(api,),
        )
        for suffix, symbol, api, family, matcher in (
            (
                "check",
                "check_yaml_update",
                "yaml_check_update",
                "values",
                matches_check,
            ),
            (
                "data-check",
                "check_yaml_data_update",
                "yaml_data_check_update",
                "values",
                matches_check,
            ),
            (
                "apply",
                "apply_yaml_update_with_decision",
                "yaml_apply_update",
                "durable-effects",
                matches_apply,
            ),
            (
                "data-apply",
                "apply_yaml_data_update_with_decision",
                "yaml_data_apply_update",
                "durable-effects",
                matches_apply,
            ),
            (
                "rollback",
                "rollback_yaml_update",
                "yaml_rollback_update",
                "durable-effects",
                matches_rollback,
            ),
            (
                "data-rollback",
                "rollback_yaml_data_update",
                "yaml_data_rollback_update",
                "durable-effects",
                matches_rollback,
            ),
        )
    ),
)

File: business-logic/classic-scanlog-core/tests/semantic_conformance/yaml_update_operations.rs
//! YAML update operations with local manifest transport and owned cache generations.
use super::{RunnerResult, invalid, text};
use classic_settings_core::SchemaCompat;
use classic_shared_core::get_runtime;
use classic_update_core::yaml_update::{
    apply_yaml_data_update_with_decision, check_yaml_data_update, rollback_yaml_data_update,
};
use classic_update_core::{
    ApprovedUpdate, ClientSchemaSet, GithubClient, RollbackOutcome, UpdateCheckConfig, UpdateError,
    YamlUpdateReport, YamlUpdateStatus, apply_yaml_update_with_decision, check_yaml_update,
    rollback_yaml_update,
};
use serde_json::{Value, json};
use std::{
    fs,
    path::{Component, Path, PathBuf},
};

/// Restore cwd before the temporary installation is removed.
struct CwdGuard(PathBuf);
impl Drop for CwdGuard {
    /// Cleanup must not replace the primary conformance failure.
    fn drop(&mut self) {
        let _ = std::env::set_current_dir(&self.0);
    }
}

/// Preserve exact files, normalizing only the parsed cache manifest's JSON ordering.
fn files(root: &Path, directory: &Path, out: &mut Vec<Value>) -> RunnerResult<()> {
    for entry in fs::read_dir(directory)? {
        let entry = entry?;
        if entry.file_type()?.is_symlink() {
            return Err(invalid("unexpected YAML update link").into());
        }
        let path = entry.path();
        if path.is_dir() {
            files(root, &path, out)?;
        } else {
            let relative = path
                .strip_prefix(root)?
                .to_string_lossy()
                .replace('\\', "/");
            let bytes = fs::read(&path)?;
            out.push(if path.file_name().is_some_and(|name| name == "manifest-latest.json") { json!({"path":relative,"json":serde_json::from_slice::<Value>(&bytes)?}) } else { json!({"path":relative,"hex":bytes.iter().map(|byte|format!("{byte:02x}")).collect::<String>()}) });
        }
    }
    out.sort_by(|a, b| a["path"].as_str().cmp(&b["path"].as_str()));
    Ok(())
}

/// Recognize typed refusals; unexpected transport or filesystem failures fail the runner.
fn error_kind(error: UpdateError) -> RunnerResult<&'static str> {
    match error {
        UpdateError::UpdateCheckDisabled => Ok("disabled"),
        UpdateError::DecisionStale { .. } => Ok("stale_decision"),
        UpdateError::Generic(message) if message.starts_with("approved decision malformed:") => {
            Ok("malformed_approval")
        }
        UpdateError::ManifestInvalid { reason }
            if reason.contains("not a valid plain cache-dir basename") =>
        {
            Ok("invalid_name")
        }
        UpdateError::Generic(message)
            if message == "rollback refused: Local Ignore YAML Data is user-owned" =>
        {
            Ok("local_ignore_refused")
        }
        other => Err(other.into()),
    }
}

/// Read the native disabled/available status envelope and all exposed file metadata.
fn status(api: &str, result: Result<YamlUpdateStatus, UpdateError>) -> RunnerResult<Value> {
    let mut value = json!({"api":api,"tag":0,"releaseTag":"","publishedAt":"","compatible":[],"incompatible":[],"incompatibleReasons":[],"unknownReason":"","error":null});
    match result? {
        YamlUpdateStatus::Disabled => {}
        YamlUpdateStatus::UpdateAvailable {
            manifest,
            compatible_files,
            incompatible_files,
        } => {
            if !incompatible_files.is_empty() {
                return Err(invalid("unexpected incompatible fixture file").into());
            }
            value["tag"] = json!(1);
            value["releaseTag"] = json!(manifest.release_tag);
            value["publishedAt"] = json!(manifest.published_at);
            value["compatible"] = json!(compatible_files.iter().map(|file|json!({"name":file.name,"schemaVersion":file.schema_version,"sha256":file.sha256,"sizeBytes":file.size_bytes,"downloadUrl":file.download_url})).collect::<Vec<_>>());
        }
        other => return Err(invalid(&format!("unexpected YAML check status {other:?}")).into()),
    }
    Ok(value)
}

/// Empty reviewed selections install nothing; explicit native errors remain separately tagged.
fn report(api: &str, result: Result<YamlUpdateReport, UpdateError>) -> RunnerResult<Value> {
    match result {
        Ok(value) => {
            if !value.installed.is_empty() || !value.failed.is_empty() {
                return Err(invalid("empty consent unexpectedly selected payloads").into());
            }
            Ok(json!({"api":api,"installed":[],"failed":[],"error":null}))
        }
        Err(error) => Ok(json!({"api":api,"installed":[],"failed":[],"error":error_kind(error)?})),
    }
}

/// Execute core operations with all cache fallbacks and network proxy routes owned by the scenario.
pub(super) fn execute(fixture: &Value, scenario: &Value) -> RunnerResult<Value> {
    let temporary = tempfile::tempdir()?;
    let root = temporary.path();
    let service = std::env::var("CLASSIC_CONFORMANCE_SERVICE")?;
    let port = service
        .strip_prefix("http://127.0.0.1:")
        .ok_or_else(|| invalid("service must be loopback"))?
        .parse::<u16>()?;
    if port == 0 {
        return Err(invalid("service port is zero").into());
    }
    fs::create_dir_all(root.join("cache"))?;
    fs::create_dir_all(root.join("bundled"))?;
    fs::write(root.join(".env"), b"")?;
    for (name, content) in fixture["files"]
        .as_object()
        .ok_or_else(|| invalid("files must be object"))?
    {
        if !name.starts_with("cache/")
            || name.contains('\\')
            || Path::new(name)
                .components()
                .any(|part| !matches!(part, Component::Normal(_)))
        {
            return Err(invalid("YAML update fixture escaped owned cache").into());
        }
        let path = root.join(name);
        fs::create_dir_all(
            path.parent()
                .ok_or_else(|| invalid("missing file parent"))?,
        )?;
        fs::write(path, text(content)?)?;
    }
    let cache = root.join("cache").to_string_lossy().into_owned();
    temp_env::with_vars(
        [
            ("LOCALAPPDATA", Some(cache.clone())),
            ("APPDATA", Some(cache.clone())),
            ("XDG_CACHE_HOME", Some(cache.clone())),
            ("HOME", Some(cache)),
            ("GITHUB_TOKEN", Some(String::new())),
            ("HTTP_PROXY", Some(service.clone())),
            ("HTTPS_PROXY", Some(service.clone())),
            ("ALL_PROXY", Some(service.clone())),
            ("NO_PROXY", Some("127.0.0.1,localhost".into())),
            ("http_proxy", Some(service.clone())),
            ("https_proxy", Some(service.clone())),
            ("all_proxy", Some(service.clone())),
            ("no_proxy", Some("127.0.0.1,localhost".into())),
        ],
        || -> RunnerResult<Value> {
            let _cwd = CwdGuard(std::env::current_dir()?);
            std::env::set_current_dir(root)?;
            // Empty owned .env plus a non-forwarding loopback proxy prevents production fallback traffic.
            let client = GithubClient::new("evildarkarchon", "CLASSIC-Fallout4")?;
            let pages = format!("{}/{}/pages", service, text(&scenario["id"])?);
            let mut current = ClientSchemaSet::new();
            current.insert("CLASSIC Main.yaml", SchemaCompat::new(2, 0), None);
            let config = |enabled| {
                (if enabled {
                    UpdateCheckConfig::enabled()
                } else {
                    UpdateCheckConfig::disabled()
                })
                .with_bundled_yaml_dir(root.join("bundled"))
            };
            let approved = if let Some(value) = fixture.get("approved") {
                ApprovedUpdate {
                    release_tag: text(&value["releaseTag"])?,
                    file_names: value["names"]
                        .as_array()
                        .ok_or_else(|| invalid("approved names must be array"))?
                        .iter()
                        .map(text)
                        .collect::<Result<Vec<_>, _>>()?,
                    file_sha256: value["digests"]
                        .as_array()
                        .ok_or_else(|| invalid("approved digests must be array"))?
                        .iter()
                        .map(text)
                        .collect::<Result<Vec<_>, _>>()?,
                }
            } else {
                ApprovedUpdate {
                    release_tag: String::new(),
                    file_names: Vec::new(),
                    file_sha256: Vec::new(),
                }
            };
            let mut before = Vec::new();
            files(root, root, &mut before)?;
            let mut results = Vec::new();
            match fixture["operation"]
                .as_str()
                .ok_or_else(|| invalid("operation missing"))?
            {
                "disabled" => {
                    results.push(status(
                        "yaml_check_update",
                        get_runtime().block_on(check_yaml_update(
                            &client,
                            &pages,
                            "yaml-data-v",
                            &current,
                            config(false),
                        )),
                    )?);
                    results.push(status(
                        "yaml_data_check_update",
                        get_runtime().block_on(check_yaml_data_update(&client, config(false))),
                    )?);
                    results.push(report(
                        "yaml_apply_update",
                        get_runtime().block_on(apply_yaml_update_with_decision(
                            &client,
                            &pages,
                            "yaml-data-v",
                            &current,
                            config(false),
                            &approved,
                        )),
                    )?);
                    results.push(report(
                        "yaml_data_apply_update",
                        get_runtime().block_on(apply_yaml_data_update_with_decision(
                            &client,
                            config(false),
                            &approved,
                        )),
                    )?);
                }
                "controlled-consent" => {
                    results.push(status(
                        "yaml_check_update",
                        get_runtime().block_on(check_yaml_update(
                            &client,
                            &pages,
                            "yaml-data-v",
                            &current,
                            config(true),
                        )),
                    )?);
                    results.push(report(
                        "yaml_apply_update",
                        get_runtime().block_on(apply_yaml_update_with_decision(
                            &client,
                            &pages,
                            "yaml-data-v",
                            &current,
                            config(true),
                            &approved,
                        )),
                    )?);
                }
                "malformed-approval" | "stale-decision" => results.push(report(
                    "yaml_apply_update",
                    get_runtime().block_on(apply_yaml_update_with_decision(
                        &client,
                        &pages,
                        "yaml-data-v",
                        &current,
                        config(true),
                        &approved,
                    )),
                )?),
                "rollback" => {
                    let name = text(&fixture["fileName"])?;
                    results.push(match rollback_yaml_update(&name) {
                        Ok(RollbackOutcome::RolledBack { file_name }) => json!({"api":"yaml_rollback_update","fileName":file_name,"rolledBack":true,"error":null}),
                        Ok(RollbackOutcome::NoPreviousVersion { file_name }) => json!({"api":"yaml_rollback_update","fileName":file_name,"rolledBack":false,"error":null}),
                        Err(error) => json!({"api":"yaml_rollback_update","fileName":name,"rolledBack":false,"error":error_kind(error)?})
                    });
                }
                "rollback-bulk" => {
                    let mut rolled = Vec::new();
                    let mut absent = Vec::new();
                    for (_, outcome) in get_runtime().block_on(rollback_yaml_data_update()) {
                        match outcome? {
                            RollbackOutcome::RolledBack { file_name } => rolled.push(file_name),
                            RollbackOutcome::NoPreviousVersion { file_name } => {
                                absent.push(file_name)
                            }
                        }
                    }
                    results.push(json!({"api":"yaml_data_rollback_update","rolledBack":rolled,"noPrevious":absent,"failedFiles":[],"failureReasons":[]}));
                }
                _ => return Err(invalid("unsupported YAML update operation").into()),
            }
            let mut after = Vec::new();
            files(root, root, &mut after)?;
            Ok(json!({"results":results,"beforeFiles":before,"files":after}))
        },
    )
}
